Feature/uiview cells

usatie opened the pull request:
closes #1 

`TableViewCell`, `TableViewHeaderFooterView`, `CollectionViewCell`, `CollectionReusableView` are available for both UIViewController and UIView.
Also, they will be internal class, so the way of registering/dequeueing is changed.

Registering UIViewController
```
ViewController.register(to: tableView)
ViewController.registerAsHeaderFooter(to: tableView)
ViewController.register(to: collectionView!)
ViewController.register(to: collectionView!, for: .header)
```

Registering UIView
```
View.register(to: tableView)
View.registerAsHeaderFooter(to: tableView)
View.register(to: collectionView!)
View.register(to: collectionView!, for: .header)
```

Dequeueing Cell with UIViewController
```
ViewController.dequeueCell(from: tableView, for: indexPath, input: input, parentViewController: self)
ViewController.dequeueHeaderFooterView(from: tableView, input: input, parentViewController: self)
ViewController.dequeueReusableCell(from: collectionView, for: indexPath, input: input, parentViewController: self)
ViewController.dequeueReusableSupplementaryView(from: collectionView, of: kind, for: indexPath, input: input, parentViewController: self)
```

Dequeueing Cell with UIView
```
View.dequeueCell(from: tableView, for: indexPath, input: input, parentViewController: self)
View.dequeueHeaderFooterView(from: tableView, input: input, parentViewController: self)
View.dequeueReusableCell(from: collectionView, for: indexPath, input: input, parentViewController: self)
View.dequeueReusableSupplementaryView(from: collectionView, of: kind, for: indexPath, input: input, parentViewController: self)
```

usatie committed:
:recycle: Add Cells+UIView, Cells+UIViewController

usatie committed:
:books: Add deprecated attributes

usatie committed:
:books: Modify documentation for CollectionReusableView

usatie committed:
:recycle: Refactored deprecated methods in MewTests

tarunon commented on Sources/Mew/Cells+UIKit/CollectionViewCell+UIKit.swift:
```diff
@@ -0,0 +1,87 @@
+    ///   - sizeConstraint: Requirement maximum size of Cell.
+    ///   - parentViewController: ParentViewController that must has collectionView.
+    /// - Returns: The Cell instance that added the ViewController.view, and the ViewController have injected dependency, VC hierarchy.
+    static func dequeueReusableCell<V>(from collectionView: UICollectionView, for indexPath: IndexPath, input: Self.Input, sizeConstraint: SizeConstraint? = nil, parentViewController: V) -> UICollectionViewCell where V: UIViewController, V: Instantiatable, Self.Environment == V.Environment {
```
命名ですが、dequeueReusableCellだとReusableは自明なので、どちらかというとdequeueCollectionViewCellのほうが良さそうですね。tableView側にも同じ実装があるので補完で混乱しなくなります。

usatie replied:
確かに、名前はどうしようか迷ってました！
いい機会なので補完で迷うものを以下のように整理してみました。
```
1. CollectionView vs TableView
2. Cell vs HeaderFooter
3. Injectable vs Injectable/Emittable
```

`1. CollectionView vs TableView`については第一引数により自明な気もします。
他についてある程度統一的にできるように、tableViewとcollectionViewどちらについてもこの名前とするのが `2. Cell vs HeaderFooter`については完全に解決できてわかりやすいかなと思いました。
```
dequeueCell(from:)
dequeueHeaderFooterView(from:)
```

`3. Injectable vs Injectable/Emittable` については、現状補完時に表示されるドキュメンテーション頼みになっていますが、現状わかりづらいですが、解決策が思い浮かんでません。

<img width="1079" alt="Screen Shot 2019-08-29 at 13 46 55" src="https://user-images.githubusercontent.com/7609060/63910921-7ed56c80-ca63-11e9-9e13-80f3e58d3810.png">

usatie replied:
コメンとはこんな感じです。あ、タイポも見つけてしまった。

<img width="1077" alt="Screen Shot 2019-08-29 at 13 53 59" src="https://user-images.githubusercontent.com/7609060/63911206-7a5d8380-ca64-11e9-8c0b-e22e03e520ca.png">

tarunon replied:
1. CollectionView vs TableView
引数で自明かどうか、というより単純に補完に4つ出てきちゃうのが開発体験悪いので、コード補完が楽になる方が良いだろうということです。
HeaderFooter/Reusableは別に付けなくてもそれぞれ自明なのでそのままで構わないかなとは思いますが、Cellは全部出てきちゃうのはあんまり嬉しくないですね。

3. Injectable vs Injectable/Emittable
これはもうしょうがないのでこのままで良いと思います。

tarunon commented on Sources/Mew/Cells+UIKit/TableViewHeaderFooterView+UIKit.swift:
```diff
@@ -0,0 +1,80 @@
+    /// Register dequeueable header/footer class for tableView
+    ///
+    /// - Parameter tableView: Parent tableView
+    static func registerAsHeaderFooter(to tableView: UITableView) {
```
これも他のメソッドをregisterAsCellとして統一したほうがいいですね、asでいいのかな?

usatie replied:
Asはなんとなく良さそうです。
```
〔…として〕登録する、〈…を〉〔…として〕登録する
```
https://ejje.weblio.jp/content/register+as


むしろ、 `to tableView`のところをwithかonなどに変えたが良さそうです。
```
register = 登録, and usually takes "with", "on", or "for".
It depends on whether the use is transitive or intransitive (他動詞か自動詞)

◯　register with xxx  
 例 : "If you register (your information) with our company, you can be assured that it will be kept safe."  (A社に登録)

◯　register on xxx
 例 : "I registered (for a class) on their website."  (website で～～の登録を)

◯　register in.  (他動詞みたいに扱う)
例　"I registered (for ～～) in March."  (３月に(なんとかの)登録をしました

◯　register for～～ (～～の登録をする)
例 "I registered for the test"

◯　register to～～ ((【to 動詞】～～をするのを登録する)
例 "I registered to take the test"
```
https://hinative.com/ja/questions/234384

usatie replied:
Asについてはdequeueの時に省略してるので、省略してもいいかもですね。
どちらかに統一したほうがいい気がしました。

Asなし
```
registerCell(on:)
registerHeaderFooter(on:)
dequeueCell(from:)
dequeueHeaderFooter(from:)
```

Asあり
```
registerAsCell(on:)
registerAsHeaderFooter(on:)
dequeueAsCell(from:)
dequeueAsHeaderFooter(from:)
```

tarunon commented on Sources/Mew/Cells+UIKit/ViewController.swift:
```diff
@@ -0,0 +1,41 @@
+
+/// ViewController wrapper for View to use UIView as T of TableViewCell.
+/// T should be `UIView & Injecatble & Instantiatable`
+internal class ViewController<T: UIView>: UIViewController, Instantiatable where T: Instantiatable {
```
UIViewに関しては、Instantiatableで縛ると使いにくくなってしまうケースがある気がしています。
`.init(frame: .zero)`を使うようにしても良いかもしれません。
現状これで問題なさそうならこれでも良いと思います。

usatie replied:
確かにそうですね。やってみます！

usatie replied:
ViewControllerはInstantiabaleじゃないと、天下り的にこれまでの`CollectionViewCell.dequeued()`や`TableViewCell.dequeued()`が使えないのですが、InstantiatableのEnvironmentはどこから確定できるようにするのがいいでしょうか？

`ViewController: UIViewController, Instantiatable, Injectable`
`T: UIView, Injectable`

usatie replied:
このままViewControllerをInstantiatableにできれば綺麗なんですが・・・
```
internal class ViewController<T: UIView>: UIViewController {
    var content: T = T(frame: .zero)
}

extension ViewController: Injectable where T: Injectable {
    typealias Input = T.Input
    func input(_ input: T.Input) {
        content.input(input)
    }
}
```

tarunon replied:
ViewController自体はInstantiatableで良くって、中に登録するViewは.init(frame: CGFrame)さえ制約にあれば良いのでは、ということでした。或いは`.init(with aCoder:NSCoder)`を使っても良いのかも(可能なら)

usatie replied:
ViewをInjectableのみにすると、ViewControllerがInstantiatableになるために必要なEnvironmentをどこから取ってくるか、という問題が発生して困っています。
下記のようにViewControllerを二つの型からなるGenericsにすると、一応できるのですが、そうするとregisterの時に余計な引数を渡す必要が出てきます。

```
/// ViewController wrapper for View to use UIView as T of TableViewCell.
/// T should be `UIView & Injecatble`
internal class ViewController<T: UIView, Parent: Instantiatable>: UIViewController, Instantiatable  where T: Injectable {
    // MARK: - Instantiatable
    typealias Input = T.Input
    typealias Environment = Parent.Environment
    var environment: Environment
    var content: T = T(frame: .zero)
    required init(with input: Input, environment: Environment) {
        self.content.input(input)
        self.environment = environment
        super.init(nibName: nil, bundle: nil)
    }
    
    @available(*, unavailable)
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
}

extension ViewController: Injectable {
    func input(_ input: T.Input) {
        content.input(input)
    }
}
```

![Screen Shot 2019-08-31 at 14 55 20](https://user-images.githubusercontent.com/7609060/64059875-626e3700-cbff-11e9-86f7-e1b41394c1be.png)
>>>
```
// MARK: - CollectionViewCell with UIView
public extension Instantiatable where Self: UIView, Self: Injectable {
    /// Register dequeueable cell class for collectionView
    ///
    /// - Parameter collectionView: Parent collectionView
    static func registerAsCollectionViewCell(on collectionView: UICollectionView) {
        ViewController<Self>.registerAsCollectionViewCell(on: collectionView)
    }
    
    /// Dequeue Injectable cell instance from collectionView
    ///
    /// - Parameters:
    ///   - collectionView: Parent collectionView that must have registered cell.
    ///   - indexPath: indexPath for dequeue.
    ///   - input: The ViewController's input.
    ///   - sizeConstraint: Requirement maximum size of Cell.
    ///   - parentViewController: ParentViewController that must has collectionView.
    /// - Returns: The Cell instance that added the ViewController.view, and the ViewController have injected dependency, VC hierarchy.
    static func dequeueAsCollectionViewCell<V>(from collectionView: UICollectionView, for indexPath: IndexPath, input: Self.Input, sizeConstraint: SizeConstraint? = nil, parentViewController: V) -> UICollectionViewCell where V: UIViewController, V: Instantiatable, Self.Environment == V.Environment {
        return ViewController<Self, V>.dequeueAsCollectionViewCell(from: collectionView, for: indexPath, input: input, sizeConstraint: sizeConstraint, parentViewController: parentViewController)
    }
}
```

tarunon replied:
Slackでも返したように、registerのときは型パラメータのみ与えれば問題は解決しそうに思います。

usatie committed:
:pencil2: Fix typo

usatie committed:
:recycle: Rename register/dequeue functions

- registerCell(on:)
- registerHeaderFooterView(on:)
- dequeueCell(from:)
- dequeueHeaderFooterView(from:)

usatie commented:
register/dequeueのためのfuncの命名は、最新の状態では下記のようになっています。
```
- registerCell(on:)
- registerHeaderFooterView(on:)
- dequeueCell(from:)
- dequeueHeaderFooterView(from:)
```

usatie committed:
:recycle: Rename dequeue functions for UIView and UIViewController

- dequeueAsCollectionViewCell
- dequeueAsCollectionViewHeaderFooterView
- dequeueAsTableViewCell
- dequeueAsTableViewHeaderFooterView

usatie commented:
egister/dequeueのためのfuncの命名は、最新の状態では下記のようになっています。

```
- registerCell(on:)
- registerHeaderFooterView(on:)
```

```
- dequeueAsCollectionViewCell()
- dequeueAsCollectionViewHeaderFooterView()
- dequeueAsTableViewCell()
- dequeueAsTableViewHeaderFooterView()
```

usatie committed:
:recycle: Rename register functions

- registerAsCollectionViewCell(on:)
- registerAsCollectionViewHeaderFooterView(on:)
- registerAsTableViewCell(on:)
- registerAsTableViewHeaderFooterView(on:)

usatie commented:
register/dequeueのためのfuncの命名は、最新の状態では下記のようになっています。
```
- registerAsCollectionViewCell(on:)
- registerAsCollectionViewHeaderFooterView(on:)
- registerAsTableViewCell(on:)
- registerAsTableViewHeaderFooterView(on:)
```

```
- dequeueAsCollectionViewCell(from:)
- dequeueAsCollectionViewHeaderFooterView(from:)
- dequeueAsTableViewCell(from:)
- dequeueAsTableViewHeaderFooterView(from:)
```

tarunon commented:
統一されていれば何でも良いとは思いますがw
registerのときの型の問題は解決した状態になりましたでしょうか?

usatie committed:
:recycle: Modify ViewController to accept non Instantiatable UIView

usatie committed:
Add tests for UIView support for TableViewCell and CollectionViewCell

- Stop using deprecated method in tests

usatie commented:
@tarunon 
registerの時には、下記のように型をパラメーターとして渡す形にしました。
```
    static func registerAsCollectionViewCell<Parent: UIViewController>(on collectionView: UICollectionView, parent: Parent.Type) where Parent: Instantiatable {
        ViewController<Self, Parent>.registerAsCollectionViewCell(on: collectionView)
    }
```

tarunon commented on Sources/Mew/Cells+UIKit/ViewController.swift:
```diff
@@ -0,0 +1,41 @@
+
+/// ViewController wrapper for View to use UIView as T of TableViewCell.
+/// T should be `UIView & Injecatble & Instantiatable`
+internal class ViewController<T: UIView, Parent: UIViewController>: UIViewController, Instantiatable where Parent: Instantiatable, T: Injectable {
```
多分Parentの型は不要で、Environmentさえ解れば良いので、型パラを渡すときにEnvironmentを渡すようにした方が良さそうです。

tarunon commented:
@tttakaoka さんには確認したのですが、私が直接コミットしても良いでしょうか?
最後体裁整えるところだけだと思うので、ぱぱっとやってしまおうと思います。

usatie commented:
@tarunon もちろんです！

tarunon committed:
Refactoring

tarunon committed:
add accurancy for CI

tarunon commented:
アクセス権付与していただけますと 🙏

usatie commented:
@tarunon 
アクセス権は実は随分前にご招待していました！
https://github.com/yenom/Mew/invitations

tarunon commented:
失礼しました 🙇 
一旦公開APIとしては十分な形になったと思います、一部Warningが出てるのが気になりますが、後でも治せるかと思います。

usatie reviewed (commented):
LGTMです！
